File: Bronze/2016/February/02_circular_barn/CircularBarn.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.PrintWriter;

/**
 * @author Red
 * @version 1.0
 */
public class CircularBarn {
    public static void main(String[] args) throws Exception {
        BufferedReader r = new BufferedReader(new FileReader("cbarn.in"));
        PrintWriter pw = new PrintWriter("cbarn.out");
        int n = Integer.parseInt(r.readLine());
        // 记录每个房间牛的数量 以及牛的总数
        int[] rooms = new int[n];
        int cowCount = 0;
        for (int i = 0; i < n; i++) {
            rooms[i] = Integer.parseInt(r.readLine());
            cowCount += rooms[i];
        }
        // 从零号房间进入的总距离
        int lastDistance = 0;
        for (int i = 1; i < n; i++) {
            lastDistance += i * rooms[i];
        }
        // 依次计算从下一个房间进入的总距离
        int minDistance = lastDistance;
        for (int i = 1; i < n; i++) {
            // 当入口从i改为i+1时，相对于i为入口时的距离，i号房间的牛要多走一圈，其他牛都少走1步
            int curDistance = lastDistance - (cowCount - rooms[i - 1]) + (n - 1) * rooms[i - 1];
            if (curDistance < minDistance) {
                minDistance = curDistance;
            }
            lastDistance = curDistance;
        }
        pw.println(minDistance);
        pw.close();
    }
}

File: Bronze/2016/February/02_circular_barn/circular_barn.cpp
#include <iostream>
#include <fstream>
#include <vector>

using namespace std;

int main() {
    freopen("cbarn.in", "r", stdin);
    freopen("cbarn.out", "w", stdout);

    int n;
    cin >> n;

    // 记录每个房间牛的数量以及牛的总数
    vector<int> rooms(n);
    int cow_count = 0;
    for (int i = 0; i < n; i++) {
        cin >> rooms[i];
        cow_count += rooms[i];
    }

    // 从零号房间进入的总距离
    int last_distance = 0;
    for (int i = 1; i < n; i++) {
        last_distance += i * rooms[i];
    }

    // 依次计算从下一个房间进入的总距离
    int min_distance = last_distance;
    for (int i = 1; i < n; i++) {
        // 当入口从 i 改为 i+1 时，相对于 i 为入口时的距离，
        // i 号房间的牛要多走一圈，其他牛都少走 1 步
        int cur_distance = last_distance - (cow_count - rooms[i - 1]) + (n - 1) * rooms[i - 1];
        if (cur_distance < min_distance) {
            min_distance = cur_distance;
        }
        last_distance = cur_distance;
    }

    cout << min_distance << endl;

    return 0;
}
